### ISSUE TITLE
No warning on shifted boolean expression

### ISSUE THREAD

**Issue by arsenm** [llvm/llvm-project#28334]
|  |  |
| --- | --- |
| Bugzilla Link | [27960](https://llvm.org/bz27960) |
| Version | unspecified |
| OS | All |
| CC | @hfinkel,@zygoloid,@Weverything,@seanm |

## Extended Description 
Code such as this produces no warning

```cpp
void bar(int x)
{
    if ((x == 0) >> 3)
    {

    }
}
```

Discovered from problem fixed in r271356

**Comment by angelomatnigoogle** [comment]
I'm interested in working on this issue despite being unfamiliar with development on clang diagnostics. I'll try to get up to speed on that in the coming days.

Conceptually, I agree that a warning that suggests "b >> x", for b of type bool, be re-written as "b & !x" prevents unintended bugs like in [r271356](https://reviews.llvm.org/D20845). I think that "b << x" is tricky but should be tackled because shifting a boolean is probably done unintentionally 99% of the time. Would warning on "b << x" with suggested re-write "b & (x | 31)" seem logical? Is there a cleaner way to express the assumed bool to int conversion?

**Comment by shafik** [comment]
I think `DiagnoseBadShiftValues(...)` might be where we want to diagnose this since we are catching other such bad shift values there. I think you would need to do something like `LHS->get()->IgnoreImpCasts()->getType()->isBooleanType()` or maybe `IgnoreParenImpCasts()`.

**Comment by angelomatni** [comment]
Will figure out how to properly link this to this issue later, but here's my work in progress. Adding the check to DiagnoseBadShiftValues caused it not to be reached in the unit test I wrote up; I will figure out why later.
https://reviews.llvm.org/D141414

**Comment by arsenm** [comment]
> Will figure out how to properly link this to this issue later, but here's my work in progress. Adding the check to DiagnoseBadShiftValues caused it not to be reached in the unit test I wrote up; I will figure out why later.
> https://reviews.llvm.org/D141414

I think You just put "Fixes #28334" in the commit message. It at least works when you push it, don't think Phabricator is scraped for it

**Comment by arsenm** [comment]
ping? This patch was never pushed and didn't make the move to a PR

**Comment by shafik** [comment]
@angelomatni if you are not able to finish this up I will label this a good first issue and let someone take a shot at it. Please let me know.

**Comment by angelomatni** [comment]
Yes, at this point I think it is better to label this a good first issue; I have not found the bandwidth to prioritize learning how to contribute to llvm
